### ISSUE TITLE
Ability to localize all DataAnnotations without specifiing ErrorMessage

### ISSUE THREAD

**Issue by tbolon** [aspnet/Mvc#6696]
Currently, [ValidationAttributeAdapterOfTAttribute.GetErrorMessage](https://github.com/aspnet/Mvc/blob/900a5c7c4c789e71bf18e8e90bf9b28b4e51013f/src/Microsoft.AspNetCore.Mvc.DataAnnotations/ValidationAttributeAdapterOfTAttribute.cs#L67) uses IStringLocalizer only when ErrorMessage is set:
```csharp
protected virtual string GetErrorMessage(ModelMetadata modelMetadata, params object[] arguments)
{
    if (modelMetadata == null)
    {
        throw new ArgumentNullException(nameof(modelMetadata));
    }

    if (_stringLocalizer != null &&
        !string.IsNullOrEmpty(Attribute.ErrorMessage) &&
        string.IsNullOrEmpty(Attribute.ErrorMessageResourceName) &&
        Attribute.ErrorMessageResourceType == null)
    {
        return _stringLocalizer[Attribute.ErrorMessage, arguments];
    }

    return Attribute.FormatErrorMessage(modelMetadata.GetDisplayName());
}
```

The consequence is that you have to set the ErrorMessage property each time you want to translate an error message.

Suppose you just want to translate the default RequiredAttribute ErrorMessageString, which is `The {0} field is required.` for all your Model classes.

1) You must override the default DataAnnotationLocalizerProvider to return a shared resource.
2) You add a generic entry named, for example, "DataAnnotations_Required" with the translated value format: `Le champ {0} doit être renseigné.`
3) You must replace all the `[Required]` attributes with `[Required(ErrorMessage = "DataAnnotations_Required")]`

More generally, there is no way to just adapt the generic DataAnnotation validation messages to your language. (the ones in System.ComponentModel.Annotations.SR.resources) without having to replace all your data annotations.

There is no easy workaround either:

- You can't inherit from RequiredAttribute to fix the ErrorMessage property, since the framework uses strict type comparison
- Replacing the default ValidationAttributeAdapterProvider is not trivial, because you have to replace all adapters with custom ones to override the GetErrorMessage method.

Is there a better way to achieve localization for all default data annotation error messages ? Or room for improvement in asp.net core mvc to reduce the burden of writing all this custom code ?

**Comment by Eilon** [comment]
@tbolon unfortunately the problem is that the built-in data annotations will return already-localized resources if the right resources are installed on the system (via NuGet, or in .NET Framework). Because of that, ASP.NET Core's localization system won't know what to translate off of because it will get "random" text depending on what's installed. That's why ASP.NET Core requires that you specify a specific error message so that it knows exactly what to key off of.

**Comment by tbolon** [comment]
You are right. In my case there was no satellite resource assembly loaded, so resources were always returned in English. Despite that, perhaps the [ValidationAttributeAdapterProvider](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.DataAnnotations/ValidationAttributeAdapterProvider.cs) to, at least, match sub-classes. So we could at least create a custom `LocRequiredAttribute : RequiredAttribute` class with a fixed ErrorMessage set.

One additional point to take into consideration: some validation messages are obtained through `DefaultModelBindingMessageProvider`, and there is an opportunity to translate them using MvcOptions and `Set...Accessor()` methods. eg. `SetValueMustBeANumberAccessor` used to display validation error for numbers.

This part is never addressed in the [documentation](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization#dataannotations-localization) and I discovered it only by chance.

It seems there are some inconsistencies or, at least, some rough edges regarding localization & model binding in asp.net core for now. I don't know how exactly they could be solved...

**Comment by Eilon** [comment]
Crazy idea time:

If the system sees that `ErrorMessage` isn't set, it can try to call into the loc system with some well-known key (such as `System.ComponentModel.DataAnnotations.Required.Message`) to see if the loc system returns a value not equal to that key. If it returns the key, clearly it has not been localized and so it should fall back to getting the data annotation attribute's intrinsic error message. If it returns some other value, it can assume it was localized and use that as the error message.

@DamianEdwards @ryanbrandenburg @hishamco - does this sound like it might work?

**Comment by ryanbrandenburg** [comment]
I can't think of any technical reason why that wouldn't work, but I'm kind of wary of adding magic keys which do specific things.

**Comment by mkArtakMSFT** [comment]
We've moved this issue to the Backlog milestone. This means that it is not going to happen for the coming release. We will reassess the backlog following the current release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.

**Comment by effyteva** [comment]
Any updates on this one?
It's very critical for large applications...

**Comment by annayafi** [comment]
Can this post help you? [https://geeksaddictions.wordpress.com/2016/11/28/data-annotations-fallback-on-shared-resources-if-not-found/](https://geeksaddictions.wordpress.com/2016/11/28/data-annotations-fallback-on-shared-resources-if-not-found/ )

**Comment by iamdroppy** [comment]
I agree with OP, this should be enhanced.
